File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(matrix.size() == 0) return false;
        int n = matrix.size();
        int m = matrix[0].size();
        
        int lo = 0;
        int hi = n*m - 1;
        
        while(lo <= hi){
            int mid = lo + (hi - lo)/2;
            if(matrix[mid/m][mid%m] == target) 
                return true;
            else if(matrix[mid/m][mid%m] > target)
                hi--;
            else 
                lo++;
        }
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(mat, tar) {
    let m = mat.length;
    let n = mat[0].length;
    
    let lo= 0;
    let hi=(m*n)-1;
    
    while(lo <= hi){
        let mid = lo + Math.floor((hi-lo)/2);
        let val = Math.floor(mid/n);
        
        if(mat[val][mid%n] === tar){
            return true;
        }
        
        if(mat[val][mid%n] > tar){
            hi=mid-1;
        }
        else{
            lo=mid+1;
        }
    }
    return false;
};